src/com/amazon/ion/impl/ResizingPipedInputStream.java

```diff
@@ -0,0 +1,557 @@
+package com.amazon.ion.impl;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.BufferOverflowException;
+import java.nio.ByteBuffer;
+
+/**
+ * Manages a resizing buffer for production and consumption of data within a <strong>single thread</strong>.
+ * Buffered bytes may be consumed through the InputStream interface. This provides a few benefits over using
+ * a PipedOutputStream/PipedInputStream pair in a single thread:
+ * <ol>
+ *     <li>There is no risk of deadlock. Piped streams, which are intended for producing data in one
+ *     thread and consuming it in another, will block on read when no data is available and block on write
+ *     when the buffer is full. In a single-threaded context, avoiding deadlock on read requires checking
+ *     that bytes are available before every read. Avoiding deadlock on write would require checking that
+ *     the buffer is not full before every write, but there is no built-in, publicly-accessible way of doing
+ *     this with a PipedInputStream/PipedOutputStream.</li>
+ *     <li>The buffer can grow. Piped streams use a fixed-size buffer that causes blocking when full. If
+ *     used in a single-thread, this serves as a hard limit on the amount of data that can be written without
+ *     a matching read. This can require arbitrary limits on data size to be imposed by the application. The
+ *     NonBlockingCircularInputStream imposes no such limitation, but optionally allows for a maximum buffer
+ *     size to be configured to protect against unbounded growth.</li>
+ * </ol>
+ */
+public class ResizingPipedInputStream extends InputStream {
+
+    /**
+     * Handler of notifications provided by the ResizingPipedInputStream.
+     */
+    interface NotificationConsumer {
+
+        /**
+         * Bytes have been shifted to the start of the buffer in order to make room for additional bytes
+         * to be buffered.
+         * @param leftShiftAmount the amount of the left shift (also: the pre-shift read index of the first shifted
+         *                        byte).
+         */
+        void bytesConsolidatedToStartOfBuffer(int leftShiftAmount);
+    }
+
+    /**
+     * A NotificationConsumer that does nothing.
+     */
+    private static final NotificationConsumer NO_OP_NOTIFICATION_CONSUMER = new NotificationConsumer() {
+        @Override
+        public void bytesConsolidatedToStartOfBuffer(int leftShiftAmount) {
+            // Do nothing.
+        }
+    };
+
+    /**
+     * Mask to isolate a single byte.
+     */
+    private static final int SINGLE_BYTE_MASK = 0xFF;
+
+    /**
+     * The initial size of the buffer and the number of bytes by which the size of the buffer will increase
+     * each time it grows.
```

Comment:
As noted in another comment, I think the buffer can grow by other amounts too.

Reply:
Updated the comment with "unless it must grow by a smaller amount to fit within 'maximumBufferSize'."